---
title: "Rice Cold Tolerance - QTLseqr Quick-Guide"
author: "Norman Warthmann, PBG Laboratory"
date: "6/15/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#setwd("~/Downloads/")
```

## Introduction

QTLseqr is an R package for Bulk Segregant Analysis from NGS data. QTLseqr was developed and published by Ben N. Mansfeld and Rebecca Grumet. We forked their github repository (https://github.com/bmansfeld/QTLseqr/) and made minor changes to their software to adopt it to our needs. For detailed instructions on usage and theory of the statistics please consult the vignette of the original QTLseqr package: https://github.com/bmansfeld/QTLseqr/raw/master/vignettes/QTLseqr.pdf.

**If you use QTLseqr, please cite:**

> Mansfeld B.N. and Grumet R, QTLseqr: An R package for bulk segregant
> analysis with next-generation sequencing *The Plant Genome* (2018)
> [doi:10.3835/plantgenome2018.01.0006](https://dl.sciencesocieties.org/publications/tpg/abstracts/11/2/180006)

**Example Analysis**

The data for below example analysis is drawn from a study on cold tolerance in rice by Yang, Z et al. (2013). It is the same data that Mansfeld and Grumet had used. We only slightly modified their instructions taken from their github page https://github.com/bmansfeld/QTLseqr and vignette.

>Yang Z, Huang D, Tang W, Zheng Y, Liang K, Cutler AJ, et al. (2013) Mapping of Quantitative Trait Loci Underlying Cold Tolerance in Rice Seedlings via High-Throughput Sequencing of Pooled Extremes. PLoS ONE 8(7): e68433. https://doi.org/10.1371/journal.pone.0068433

The raw sequencing data for Yang Z, et al. (2013) is available from NCBI's short read archive in BioProjet PRJNA198759:

* Run SRR834931: Rice ET pool (385 extremely tolerant individuals) 
* Run SRR834927: Rice ES pool (430 extremely sensitive individuals)

We have downloaded the raw data (fasterq-dump SRR834931 SRR834927) and performed alignment (bwa) and variant calling (freebayes) against the Nipponbare reference genome (IRGSP-1.0) with our PBGL snakemake workflow (https://github.com/pbgl/dna-proto-workflow). QTLseqr analysis on the [resulting VCF file produced by freebayes](https://bss1innov1nafa1poc1.blob.core.windows.net/sample-container/Data-for-github/wGQ-Filt-freebayes~bwa~IRGSP-1.0~both-segregant_bulks~filtered-default.vcf) can be performed as outlined below.

\newpage
# Installation

Install the PBGL version of QTLseqr from github like so:

```{r, eval=FALSE}
#install the dependencies
install.packages(c( "data.table", 
                    "modeest",
                    "locfit",
                    "dplyr", 
                    "tidyr", 
                    "vcfR", 
                    "ggplot2"), dependencies=TRUE)

# install devtools
install.packages("devtools", dependencies=TRUE)

# use devtools to install QTLseqr from github
library("devtools")
#devtools::install_github("bmansfeld/QTLseqr")
#devtools::install_github("warthmann/QTLseqr")
devtools::install_github("pbgl/QTLseqr")
```

**Note:** When installing above R-packages, you may run into missing system dependencies. On a clean Ubuntu systems these may include libssl-dev liblapack-dev libopenblas-dev libcurl4-openssl-dev libxml2-dev libudunits2-dev libmariadbclient-dev gfortran libpng-dev libjpeg-dev.
Some functions of QTLseqr (e.g., counting SNPs) are implemented in C++. Hence, the install of QTLseqr from github requires the compiler. On Linux this should work out of the box, for Windows and Mac you will need Rtools or Xcode, respectively. You might need assistance from your system administrator.

Do the install from github if you can. Alternatively, we provide binaries of the pbgl version of the QTLseqr R-package for local install: [Linux](https://bss1innov1nafa1poc1.blob.core.windows.net/sample-container/Data-for-github/pbglQTLseqr-binaries/QTLseqr_0.7.5.2_R_x86_64-pc-linux-gnu.tar.gz), 
[Mac](https://bss1innov1nafa1poc1.blob.core.windows.net/sample-container/Data-for-github/pbglQTLseqr-binaries/QTLseqr_0.7.5.2.tgz), 
[Windows](https://bss1innov1nafa1poc1.blob.core.windows.net/sample-container/Data-for-github/pbglQTLseqr-binaries/QTLseqr_0.7.5.2.zip). They were compiled with R version 4.2.


```{r, message=FALSE}
#load dependencies
library("devtools")
library("data.table")
library("dplyr")
library("tidyr")
library("vcfR")
library("ggplot2")
library("QTLseqr")
```

\newpage
## Step-by-Step Guide

**Define the input:**

```{r}
#Set samples
HighBulk <- "ET-pool-385" # SRA-run: SRR834931
LowBulk <- "ES-pool-430" # SRA-run: SRR834927 

#set file name of the VCF file to load
file <- "wGQ-Filt-freebayes~bwa~IRGSP-1.0~both-segregant_bulks~filtered-default.vcf"

#Specify which chromosomes should be included in the analysis (i.e., exclude smaller contigs)
Chroms <- c("NC_029256.1",
            "NC_029257.1",
            "NC_029258.1",
            "NC_029259.1",
            "NC_029260.1",
            "NC_029261.1",
            "NC_029262.1",
            "NC_029263.1",
            "NC_029264.1",
            "NC_029265.1",
            "NC_029266.1",
            "NC_029267.1")
```

Chromosome information can be found in the VCF file header.
Open the VCF file in a text editor and find the '##contig' lines. E.g.:

##reference=genomes_and_annotations/IRGSP-1.0/GCF_001433935.1_IRGSP-1.0_genomic.fna 
##contig=<ID=NC_029256.1,length=43270923>  
##contig=<ID=NC_029257.1,length=35937250>  
##contig=<ID=NC_029258.1,length=36413819>  
##contig=<ID=NC_029259.1,length=35502694>  
##contig=<ID=NC_029260.1,length=29958434>  
##contig=<ID=NC_029261.1,length=31248787>  
##contig=<ID=NC_029262.1,length=29697621>  
##contig=<ID=NC_029263.1,length=28443022>  
##contig=<ID=NC_029264.1,length=23012720>  
##contig=<ID=NC_029265.1,length=23207287>  
##contig=<ID=NC_029266.1,length=29021106>  
##contig=<ID=NC_029267.1,length=27531856> 


**Load the data from the VCF file:**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, cache = TRUE}
df <-
  importFromVCF(
    file = file,
    highBulk = HighBulk,
    lowBulk = LowBulk,
    chromList = Chroms
    )
```
\newpage
**Filter the Variants (user-defined criteria)**

```{r, cache=TRUE} 
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 70,
        maxTotalDepth = 200,
        minSampleDepth = 30,
        depthDifference = 100,
        minGQ = 150,
        verbose = TRUE
    )
```

\newpage
**Run the G` (G prime) Analysis**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, cache=TRUE}
df_filt <- 
    runGprimeAnalysis(
        SNPset = df_filt,
        windowSize = 1e6,
        outlierFilter = "deltaSNP",
        filterThreshold = 0.1
    )
```

**Run the Delta SNP Analysis**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, cache=TRUE}
df_filt <- 
    runQTLseqAnalysis(
        SNPset = df_filt,
        windowSize = 1e6,
        popStruc = "F2",
        bulkSize = c(385, 430), 
        replications = 10000,
        intervals = c(95, 99)
    )
```

**Plot SNP Density**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, fig.align = 'center', fig.cap="SNP Density", fig.width=9}

#Plot SNP density
plotQTLStats(SNPset = df_filt, var = "nSNPs", plotIntervals = TRUE)
```

\newpage
**Plot G'**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, fig.align = 'center', fig.cap="G' Values",  fig.width=9}
plotQTLStats( SNPset = df_filt,
              var = "Gprime",
              plotThreshold = TRUE,
              q = 0.02
              )
```

```{r, warning=FALSE, message=FALSE, error=FALSE}
#get a list of significant Loci
getQTLTable(SNPset = df_filt, method = "Gprime", alpha=0.02)
```

\newpage
**Plot Delta SNP**
```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, fig.align = 'center', fig.cap="Delta SNP"}
#Plot Delta-SNP
plotQTLStats(SNPset = df_filt, 
             var = "deltaSNP", 
             plotIntervals = TRUE)
```

```{r, warning=FALSE, message=FALSE, error=FALSE}
#export table as csv file
getQTLTable(SNPset = df_filt, method = "QTLseq", 
            export=TRUE, fileName= "my_first_BSA_result.csv")
```

\newpage
**Plot only a subset of Chromosomes**

```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, fig.align = 'center', fig.cap="G' Chromsome 1,2,5,8"}
#Plot only a subset of Chromosomes
plotQTLStats( SNPset = df_filt, 
              var = "Gprime", 
              plotThreshold = TRUE, 
              q = 0.02, 
              subset=c("NC_029256.1","NC_029257.1", "NC_029260.1", "NC_029263.1")
)
```

\newpage
```{r, results='hide', warning=FALSE, message=FALSE, error=FALSE, fig.align = 'center', fig.cap="G' Chromsome 1,2,5,8"}
#Plot only a subset of Chromosomes
plotQTLStats( SNPset = df_filt, 
              var = "deltaSNP", 
              plotIntervals = TRUE, 
              subset=c("NC_029256.1","NC_029257.1", "NC_029260.1", "NC_029263.1")
)
```

